<a href="https://colab.research.google.com/github/kennedot/Github-and-Jupyter-setup/blob/main/data_cleaning_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Cleaning

In this assignment, we're going to practice data cleaning with two datasets. Both of these require the kind of cleanup you saw in the practice notebook. But these also ask you to clean data in order to ensure the privacy of individuals. So you're going to work with a version of the OKCupid data that we looked at in this week's case study.

First, let's import some libraries. For this assignment, we'll need `pandas`.

In [1]:
# IMPORT LIBRARIES BELOW
import pandas as pd

# we're going to open an excel file...you might need this library
!pip install openpyxl

## Part 1: OkCupid
Let's start with the OKCupid dataset.

### Reading in the data
The data for this assignment is in a file named "okcupid.xlsx" in the "data" directory. You can access it at "data/okcupid.xlsx". Import the data into a dataframe named `okcupid_data`.

In [14]:
# the data is at this URL
okcupid_data_url = "https://raw.githubusercontent.com/bsmith-classroom/datasets/master/okcupid.xlsx"

# Read the excel file at okcupid_data_url into a dataframe named okcupid_data
okcupid_data = pd.read_excel(okcupid_data_url)

### Do some simple overviews of the data
The next few cells will ask you to use different `pandas` functions to learn about the dataset.

In [15]:
# Use `head()` to get an overview of the first few records in the dataset
okcupid_data.head()

age       body_type               diet    drinks      drugs  \
0   22  a little extra  strictly anything  socially      never   
1   35         average       mostly other     often  sometimes   
2   38            thin           anything  socially        NaN   
3   23            thin         vegetarian  socially        NaN   
4   29        athletic                NaN  socially      never   

                           education               essay0  \
0      working on college/university  removed_for_privacy   
1              working on space camp  removed_for_privacy   
2     graduated from masters program  removed_for_privacy   
3      working on college/university  removed_for_privacy   
4  graduated from college/university  removed_for_privacy   

                essay1               essay2               essay3  ...  \
0  removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   
1  removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   
2  removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   
3  removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   
4  removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   

                          location  \
0  south san francisco, california   
1              oakland, california   
2        san francisco, california   
3             berkeley, california   
4        san francisco, california   

                                      offspring orientation  \
0  doesn&rsquo;t have kids, but might want them    straight   
1  doesn&rsquo;t have kids, but might want them    straight   
2                                           NaN    straight   
3                       doesn&rsquo;t want kids    straight   
4                                           NaN    straight   

                        pets                                  religion sex  \
0  likes dogs and likes cats     agnosticism and very serious about it   m   
1  likes dogs and likes cats  agnosticism but not too serious about it   m   
2                   has cats                                       NaN   m   
3                 likes cats                                       NaN   m   
4  likes dogs and likes cats                                       NaN   m   

                                 sign     smokes  \
0                              gemini  sometimes   
1                              cancer         no   
2  pisces but it doesn&rsquo;t matter         no   
3                              pisces         no   
4                            aquarius         no   

                                              speaks     status  
0                                            english     single  
1  english (fluently), spanish (poorly), french (...     single  
2                               english, french, c++  available  
3                           english, german (poorly)     single  
4                                            english     single  

[5 rows x 31 columns]

In [16]:
# Check the number of rows and columns with `shape`
okcupid_data.shape

(59946, 31)

In [17]:
# Check the data types of each column with `info()`.
# Do the 'non-null' totals for each column equal the
# number of rows you found from `shape`? Why or why
# not?
okcupid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 31 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   age          59946 non-null  int64  
 1   body_type    54650 non-null  object 
 2   diet         35551 non-null  object 
 3   drinks       56961 non-null  object 
 4   drugs        45866 non-null  object 
 5   education    53318 non-null  object 
 6   essay0       59946 non-null  object 
 7   essay1       59946 non-null  object 
 8   essay2       59946 non-null  object 
 9   essay3       59946 non-null  object 
 10  essay4       59946 non-null  object 
 11  essay5       59946 non-null  object 
 12  essay6       59946 non-null  object 
 13  essay7       59946 non-null  object 
 14  essay8       59946 non-null  object 
 15  essay9       59946 non-null  object 
 16  ethnicity    54263 non-null  object 
 17  height       59940 non-null  float64
 18  income       59943 non-null  float64
 19  job 

In [18]:
# Use `describe()` to get some basic statistics.
# describe has an optional argument called 'include='.
# Set that equal to 'all', and you'll to see numerical
# and categorical variables

okcupid_data.describe(include= 'all')

age body_type             diet    drinks  drugs  \
count   59946.000000     54650            35551     56961  45866   
unique           NaN        12               18         6      3   
top              NaN   average  mostly anything  socially  never   
freq             NaN     14652            16585     41780  37724   
mean       32.340290       NaN              NaN       NaN    NaN   
std         9.452779       NaN              NaN       NaN    NaN   
min        18.000000       NaN              NaN       NaN    NaN   
25%        26.000000       NaN              NaN       NaN    NaN   
50%        30.000000       NaN              NaN       NaN    NaN   
75%        37.000000       NaN              NaN       NaN    NaN   
max       110.000000       NaN              NaN       NaN    NaN   

                                education               essay0  \
count                               53318                59946   
unique                                 32                    1   
top     graduated from college/university  removed_for_privacy   
freq                                23959                59946   
mean                                  NaN                  NaN   
std                                   NaN                  NaN   
min                                   NaN                  NaN   
25%                                   NaN                  NaN   
50%                                   NaN                  NaN   
75%                                   NaN                  NaN   
max                                   NaN                  NaN   

                     essay1               essay2               essay3  ...  \
count                 59946                59946                59946  ...   
unique                    1                    1                    1  ...   
top     removed_for_privacy  removed_for_privacy  removed_for_privacy  ...   
freq                  59946                59946                59946  ...   
mean                    NaN                  NaN                  NaN  ...   
std                     NaN                  NaN                  NaN  ...   
min                     NaN                  NaN                  NaN  ...   
25%                     NaN                  NaN                  NaN  ...   
50%                     NaN                  NaN                  NaN  ...   
75%                     NaN                  NaN                  NaN  ...   
max                     NaN                  NaN                  NaN  ...   

                         location                offspring orientation  \
count                       59943                    24383       59943   
unique                        199                       15           3   
top     san francisco, california  doesn&rsquo;t have kids    straight   
freq                        31064                     7559       51603   
mean                          NaN                      NaN         NaN   
std                           NaN                      NaN         NaN   
min                           NaN                      NaN         NaN   
25%                           NaN                      NaN         NaN   
50%                           NaN                      NaN         NaN   
75%                           NaN                      NaN         NaN   
max                           NaN                      NaN         NaN   

                             pets     religion    sex  \
count                       40023        39717  59943   
unique                         15           45      2   
top     likes dogs and likes cats  agnosticism      m   
freq                        14813         2723  35826   
mean                          NaN          NaN    NaN   
std                           NaN          NaN    NaN   
min                           NaN          NaN    NaN   
25%                           NaN          NaN    NaN   
50%                           NaN          NaN    NaN   
75%        

Using `describe()` in this way shows us the mean, standard deviation, max, min, and interquartile values for numerical data. We also see the number of unique values and the most common value for categorical data. And we see a lot of `NaN` in the categorical data. Remember that these are missing values we need to clean up before we can do any data analysis. Before we start dealing with the missing values, let's look at our columns to see if there are any we won't need.

In [19]:
# Use `columns` to see the column names
okcupid_data.columns

Index(['age', 'body_type', 'diet', 'drinks', 'drugs', 'education', 'essay0',
       'essay1', 'essay2', 'essay3', 'essay4', 'essay5', 'essay6', 'essay7',
       'essay8', 'essay9', 'ethnicity', 'height', 'income', 'job',
       'last_online', 'location', 'offspring', 'orientation', 'pets',
       'religion', 'sex', 'sign', 'smokes', 'speaks', 'status'],
      dtype='object')

### Drop some of the unnecessary columns
You should notice several columns named essay0, essay1,...essay9. The data in those columns all read `removed_for_privacy` since they originally contained information that should have been redacted. There's also a column named `last_online` that we won't need.   

ELiminating unnecssary data is just as important as filling in missing data!   

Use `drop()` with `axis='columns'` to remove the `essay#` and `last_online` columns from `okcupid_data`.

Hint: `drop()` takes a list (e.g., ['essay0', 'essay1'...]) of column names that you want to drop.

In [20]:
# drop last_online and the essay columns below
okcupid_data = okcupid_data.drop(columns=['last_online', 'essay0', 'essay1', 'essay2', 'essay3', 'essay4', 'essay5', 'essay6', 'essay7', 'essay8', 'essay9'])

### See where we're at after dropping columns
This is a good time to look at the first few rows again using `head`

In [21]:
# Check the head of okcupid_data again
okcupid_data.head()

age       body_type               diet    drinks      drugs  \
0   22  a little extra  strictly anything  socially      never   
1   35         average       mostly other     often  sometimes   
2   38            thin           anything  socially        NaN   
3   23            thin         vegetarian  socially        NaN   
4   29        athletic                NaN  socially      never   

                           education            ethnicity  height   income  \
0      working on college/university         asian, white    75.0     -1.0   
1              working on space camp                white    70.0  80000.0   
2     graduated from masters program                  NaN    68.0     -1.0   
3      working on college/university                white    71.0  20000.0   
4  graduated from college/university  asian, black, other    66.0     -1.0   

                           job                         location  \
0               transportation  south san francisco, california   
1         hospitality / travel              oakland, california   
2                          NaN        san francisco, california   
3                      student             berkeley, california   
4  artistic / musical / writer        san francisco, california   

                                      offspring orientation  \
0  doesn&rsquo;t have kids, but might want them    straight   
1  doesn&rsquo;t have kids, but might want them    straight   
2                                           NaN    straight   
3                       doesn&rsquo;t want kids    straight   
4                                           NaN    straight   

                        pets                                  religion sex  \
0  likes dogs and likes cats     agnosticism and very serious about it   m   
1  likes dogs and likes cats  agnosticism but not too serious about it   m   
2                   has cats                                       NaN   m   
3                 likes cats                                       NaN   m   
4  likes dogs and likes cats                                       NaN   m   

                                 sign     smokes  \
0                              gemini  sometimes   
1                              cancer         no   
2  pisces but it doesn&rsquo;t matter         no   
3                              pisces         no   
4                            aquarius         no   

                                              speaks     status  
0                                            english     single  
1  english (fluently), spanish (poorly), french (...     single  
2                               english, french, c++  available  
3                           english, german (poorly)     single  
4                                            english     single

### Fill in missing values with `fillna()`
The unnecessary rows should be gone, and we can focus on filling in some of the missing values. Let's start with the numerical values.  

Call `describe` below to see information about the numerical values only (remember that `include='all'` gives us info on all the variables).


In [22]:
# Call describe() on ok_data
okcupid_data.describe()

age        height          income
count  59946.000000  59940.000000    59943.000000
mean      32.340290     68.295229    20033.557880
std        9.452779      3.994891    97348.525138
min       18.000000      1.000000       -1.000000
25%       26.000000     66.000000       -1.000000
50%       30.000000     68.000000       -1.000000
75%       37.000000     71.000000       -1.000000
max      110.000000     95.000000  1000000.000000

A few things stand out:
1. The youngest (minimum) age is 18, the oldest (maximum) is 110. While it'd be great to be looking for a partner at 110, we can assume this is probably not accurate. Still, the mean is 32.3 years old which seems reasonable. Let's assume 110 is an outlier of the sort we always find in data analysis.
2. The mean height is 68.3 inches, aroud 5.7 feet tall. The smallest height is 1 inch while the tallest is 95 inches or 7.9 feet. The Guinness world record holder for height is [Robert Wadlow](https://www.guinnessworldrecords.com/records/hall-of-fame/robert-wadlow-tallest-man-ever) who was 8'11" tall. Definitely an outlier! But we have to decide what to do with missing height values.
3. `income` is interesting. Notice that the minimum and 25%, 50%, and 75% values are all -1.0. These values don't make sense, and there seem to be a lot of them. We'll have to deal with those too.

For both  height and income, lets's see how many (if any) missing values we're dealing with.

In [23]:
# Fill in code to compute the number of missing heights and incomes below
missing_heights = 0     # missing height code
missing_incomes = 0     # missing income code

print("Number of missing height values =", missing_heights)
print("Number of missing income values =", missing_incomes)


Number of missing height values = 0
Number of missing income values = 0


There aren't that many missing values. Let's make some assumptions.
1. Height is normally distributed (see [here](https://ourworldindata.org/human-height) for more). So we can fill in the missing heights with the `mean` of the heights and feel comfortable.
2. The incomes? So many of the values are equal to `-1.0` that we might not be able to do anything with this column. So let's just replace the (few) `NaN` values with the mean of the income data.

`fillna()` is a good function to replace the missing values. Do that below.

In [25]:
# first, get the mean height
mean_height = 68     ## COMPUTE THE MEAN OF HEIGHT HERE
print("The mean height is", mean_height)

# Now, replace the missing heights with the mean of the height column
okcupid_data['height'] = mean_height ## YOUR CODE HERE

# Next, get the mean income
mean_income = 20033     ## COMPUTE THE MEAN OF INCOME HERE

# Now, replace the missing incomes with the mean of the income column
okcupid_data['income'] = mean_income ## YOUR CODE HERE


The mean height is 68


You should double check these to make sure all the missing values are filled in.

In [31]:
# Do a count of the missing values for height and income here.
# They should all be removed

# count missing height values
missing_height_values = okcupid_data['height'].isnull().sum()

# count missing income values
missing_income_values = okcupid_data['income'].isnull().sum()

# The lines below will let you see your results and test to make
# sure there are no missing values
print("Missing heights =", missing_height_values)
print("Missing incomes =", missing_income_values)

assert missing_height_values == 0, "There shouldn't be any missing heights."
assert missing_income_values == 0, "There shouldn't be any missing incomes."

Missing heights = 0
Missing incomes = 0


### Body type
Now let's deal with ther body type. We saw from the `describe` output that there are `NaN` values in that column. And `body_type` is a categorical variable, so we can't just take "the mean" and fill in the results. Let's do a few quick calculations to see how bad this is.  

First, count the number of missing `body_type` values. Then add up (sum) the number of values for the column: This could give us an idea of an "average" body_type.

In [32]:
# Count the missing values for body_type
missing_body_type_values = okcupid_data.body_type.isna().sum()

print("Missing body_type values =", missing_body_type_values)

Missing body_type values = 5296


That's a lot of missing values to replace. Maybe we can get some ideas about how to replace these by looking at the values that exist.  

Use `value_counts()` to look at how people describe their `body_type`.

In [33]:
# Use value_counts() to count the unique entries in the body_type column
okcupid_data.body_type.value_counts()

body_type
average           14652
fit               12711
athletic          11819
thin               4711
curvy              3924
a little extra     2629
skinny             1777
full figured       1009
overweight          444
jacked              421
used up             355
rather not say      198
Name: count, dtype: int64

There are many more unknown values in `body_type` than we saw for `height` and `income`. We could just mark them as unknown, but the majority of people labelled themselves as "average" (do you see that in your calculations?). So let's assume the missing people are also 'average'. Fill in the missing values with 'average' below.

In [36]:
# Fill in the missing values with 'average' below
okcupid_data['body_type'] = okcupid_data['body_type'].fillna('average') # YOUR CODE HERE

# Let's check to make sure we got rid of the missing values
print("Missing body_type values =", okcupid_data['body_type'].isna().sum())

Missing body_type values = 0


### Diet
`diet` is another categorical variable. Let's get counts for missing values and the different responses.

In [37]:
# Count the missing values for diet
missing_diet_values = okcupid_data.diet.isna().sum()  # YOUR CODE WILL REPLACE 0

print("Missing diet values =", missing_diet_values)

Missing diet values = 24395


In [40]:
# value_counts() will show the diet column's unique entries and their quantities
missing_diet_values = okcupid_data.diet.value_counts()
missing_diet_values

diet
mostly anything        16585
anything                6183
strictly anything       5113
mostly vegetarian       3444
mostly other            1007
strictly vegetarian      875
vegetarian               667
strictly other           452
mostly vegan             338
other                    331
strictly vegan           228
vegan                    136
mostly kosher             86
mostly halal              48
strictly halal            18
strictly kosher           18
halal                     11
kosher                    11
Name: count, dtype: int64

Which value do people typically choose for `diet`? We could do what we did with `body_type` and just make the missing values equal to that value. It's probably a safe assumption, but we'll avoid that for this exercise. Let's not assume anything about diet since there could be reasons for adopting a diet (e.g., religious reasons, food allergies, etc). Let's replace the missing values here with 'unknown'.

In [43]:
# Fill in the missing values with 'unknown' below
okcupid_data['diet'] = okcupid_data['diet'].fillna('unknown')
# make sure it worked by printing the number of missing values
print("Missing diet values =", okcupid_data['diet'].isna().sum()) ##YOUR CODE HERE )


Missing diet values = 0


Everything left in the dataframe are the categorical data. We could take a look at each column to see how we want to deal with those missing values. But let's just do what we did with `diet`: We'll set the remaining missing values to 'unknown'.  

You can do this in different ways. One way is to use `info()` to find columns with non-null values that aren't equal to 59946, the number of rows in the DataFrame. for each of those, you could write a line of code like this:

`okcupid_data[<column_name>].fillna('unknown')`

That works, but what if we have 50 columns? 100 columns? It''d be easier to use a `for` loop to go through all of the columns, see which have missing values, and use `fillna()` to insert 'unknown'. The code would operate like this:

```
For every column in the data:
    If the column has missing values (if the sum of missing values is greater than 0):
        use fillna() to replace the missing values with 'unknown'
```
    
Give it a try...and mail the teaching staff if you have any problems. This could be tough if you're new to Python, but the solution is very much like doing it one column at a time.

In [44]:
# Let's make every remaining missing value equal 'unknown'
for column in okcupid_data.columns:
    if okcupid_data[column].isna().sum() > 0:
        okcupid_data[column] = okcupid_data[column].fillna('unknown')
# check okcupid_data with info() when you're done. The non_null values should all
# be the same and that number should equal the number of
# rows of data


okcupid_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   age          59946 non-null  int64 
 1   body_type    59946 non-null  object
 2   diet         59946 non-null  object
 3   drinks       59946 non-null  object
 4   drugs        59946 non-null  object
 5   education    59946 non-null  object
 6   ethnicity    59946 non-null  object
 7   height       59946 non-null  int64 
 8   income       59946 non-null  int64 
 9   job          59946 non-null  object
 10  location     59946 non-null  object
 11  offspring    59946 non-null  object
 12  orientation  59946 non-null  object
 13  pets         59946 non-null  object
 14  religion     59946 non-null  object
 15  sex          59946 non-null  object
 16  sign         59946 non-null  object
 17  smokes       59946 non-null  object
 18  speaks       59946 non-null  object
 19  status       59946 non-nu

### Cleaning up data types
Once you clean a column, you can assign it an appropriate data type.  

You can explore the data for yourselves to see that the `height` and `income` are always expressed as whole numbers (e.g., 80000.0, 98.0). Let's make those integer (`int`) values since we don't need to do any decimal calculations on them. Redefine them with `astype()`.


In [45]:
# convert the height and income columns to type int with astype
okcupid_data[['height', 'income']].astype(int)

height  income
0          68   20033
1          68   20033
2          68   20033
3          68   20033
4          68   20033
...       ...     ...
59941      68   20033
59942      68   20033
59943      68   20033
59944      68   20033
59945      68   20033

[59946 rows x 2 columns]

### Character/String issues
Let's look at the first few entries in the `offspring` column.

In [46]:
# Your code below
okcupid_data['offspring'].head()

0    doesn&rsquo;t have kids, but might want them
1    doesn&rsquo;t have kids, but might want them
2                                         unknown
3                         doesn&rsquo;t want kids
4                                         unknown
Name: offspring, dtype: object

You should see a strange looking string of characters:

> doesn\&rsquo;t have kids

What is this "\&rsquo;"? It turns out that "\&rsquo;" stands for "right single quote". This is an encoded version of a quotation character often used in HTML to ensure the character renders correctly in a web browser. That's great for HTML, but we're not doing web browsing. We want to get rid of that special encoding and replace it with an apostrophe/single quote.   

Let's use the `replace()` function to do this. First, here's a reminder of how it works:


In [47]:
# Using replace to change values in a string
prince_songs = "Prince wrote 'i would die four you' and 'four the tears in your eyes."

# replace 'four' with '4', Prince style.
prince_songs.replace('four', '4')

"Prince wrote 'i would die 4 you' and '4 the tears in your eyes."

We can do the same thing to replace "doesn\&rsquo;t;" with "doesn't" in our `offspring` column. Hint: You need to make sure you're calling `replace()` on a `str` version of the column. Here's how this looks:

In [48]:
# Replace "doesn&rsquo;t;" with "doesn't" below
okcupid_data['offspring'] = okcupid_data['offspring'].str.replace("doesn&rsquo;t", "doesn't")

# let's make sure the replace worked
okcupid_data['offspring'].head()

0    doesn't have kids, but might want them
1    doesn't have kids, but might want them
2                                   unknown
3                         doesn't want kids
4                                   unknown
Name: offspring, dtype: object

The `sign` column also has "doesn\&rsquo;t;" in it. This time, you write the code to replace those like we did the ones in `offspring`.

In [49]:
# Replace "doesn&rsquo;t;" with "doesn't" in the sign column below
okcupid_data['sign'] = okcupid_data['sign'].str.replace("doesn&rsquo;t", "doesn't")
# let's make sure the replace worked
okcupid_data['sign'].head()

0                          gemini
1                          cancer
2    pisces but it doesn't matter
3                          pisces
4                        aquarius
Name: sign, dtype: object

### Data cleaned!
Use `info()` one final time to see what you did.

In [50]:
# call info on okcupid_data below
okcupid_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   age          59946 non-null  int64 
 1   body_type    59946 non-null  object
 2   diet         59946 non-null  object
 3   drinks       59946 non-null  object
 4   drugs        59946 non-null  object
 5   education    59946 non-null  object
 6   ethnicity    59946 non-null  object
 7   height       59946 non-null  int64 
 8   income       59946 non-null  int64 
 9   job          59946 non-null  object
 10  location     59946 non-null  object
 11  offspring    59946 non-null  object
 12  orientation  59946 non-null  object
 13  pets         59946 non-null  object
 14  religion     59946 non-null  object
 15  sex          59946 non-null  object
 16  sign         59946 non-null  object
 17  smokes       59946 non-null  object
 18  speaks       59946 non-null  object
 19  status       59946 non-nu

1. There should be 20 columns: You successfully eliminated 10 early in the exercise.
2. The numerical data are all integer types now.
3. All the missing/`NaN` values are gone!

The final step: Save your data frame as a .csv file named okcupid_data.csv.

In [52]:
# Save the data to "okcupid_data.csv"
save_path = "okcupid_data.csv"
okcupid_data.to_csv(save_path)